tenseal/cpp/tensors/ckkstensor.cpp

```diff
@@ -76,7 +76,38 @@ shared_ptr<CKKSTensor> CKKSTensor::power_inplace(unsigned int power) {
 
 shared_ptr<CKKSTensor> CKKSTensor::add_inplace(
     const shared_ptr<CKKSTensor>& to_add) {
-    // TODO
+    // TODO implement broadcasting
+    if (this->_shape != to_add->_shape) {
+        // TODO provide a better message (what are the shapes)
+        throw invalid_argument("Operands doesn't have the same shape");
+    }
+
+    size_t n_jobs = this->tenseal_context()->dispatcher_size();
+
+    auto worker_func = [&](vector<Ciphertext>* dst, vector<Ciphertext>* src,
+                           size_t start, size_t end) -> bool {
+        for (size_t i = start; i < end; i++) {
+            this->tenseal_context()->evaluator->add_inplace((*dst)[i],
```

Comment:
I don;t think this is thread safe, to access dst this way.

You either use a mutex over dst/this->data, or return the batch result, and let the main thread do the computations over this->data.

Reply:
I thought writing to separate locations should be fine, but it seems the object itself isn't thread safe, thanks for spotting that.

Reply:
Aah, my bad.

Yes, that is thread safe, as long the dst object is allocated, and you won't write in the same slot from multiple threads.

But I'm a bit confused:
 - you are creating the lambda with & ref, so you can access this->data directly.
 - same for to_add, you are creating a ref to the parameters as well.

I was confused by the pointer parameters, I though you reallocate something

Reply:
Oh yeah! Initially I used pointers then I started using the lambda with & ref to access the evaluator, but now I can remove the dst and src parameters as I can access them directly.